pennylane/operation.py

```diff
@@ -573,6 +573,19 @@ def has_matrix(cls):
         """
         return cls.compute_matrix != Operator.compute_matrix
 
+    # pylint: disable=no-self-argument, comparison-with-callable
+    @classproperty
+    def has_decomposition(cls):
+        r"""Bool: Whether or not the Operator returns a defined decomposition.
+
+        Note: Child classes may have this as an instance property instead of as a class property.
+        """
+        # Some operators will overwrite `decomposition` instead of `compute_decomposition`.
+        return (
+            cls.compute_decomposition != Operator.compute_decomposition
+            or cls.decomposition != Operator.decomposition
+        )
+
```

Comment:
Remove these `has_decomposition` things, and I'll approve.

Reply:
What is wrong with providing the users a way to check whether an op can be decomposed?

Reply:
@cvjjm To my understanding, we will add them in #2986 .
@albi3ro I thought we want to do #2986 first, and then use `has_decomposition` here. If we don't a considerable part of this PR's impact is reverted.

Reply:
@cvjjm Adding `has_decomposition` will be a change to a core data structure, and deserves an independent PR.  In addition to explaining, documenting, and testing the change to `Operator`, we will need to add the attribute to all the arithmetic operations and update all the uses of `try: except DecompositionUndefinedError`.

My idea is that we have one PR that has the small incremental changes from this PR, like copying over properties instead of recalculating them, rewriting `Wires.all_wires`, and reordering methods. That could be approved and merged right now. 

Then we add `has_decomposition`. One more PR.

Then we perform further optimizations to `expand_tape`.

This makes it easier to test, track where changes came from, and review.

Reply:
I fully agree that the `has_decomposition` PR is a separate one, I just thought it might be nice to combine the `Wires.all_wires` and copying of properties with the simple addition in `Operator.expand`, which would be possible if we first finish #2986.
I don't mind opening a third PR where we do it, just thought they pair nicely :) 
Will revert `has_decomposition` stuff here, then.